library/firewall_lib_facts.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2016,2017,2020,2021 Red Hat, Inc.
+# Reusing some firewalld code
+# Authors:
+# Brennan Paciorek <bpaciore@redhat.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+module: firewall_lib_facts
+short_description: module for firewall role
+requirements:
+  - python3-firewall or python-firewall
+description:
+  Get firewalld settings, including default settings
+  and custom settings.
+options:
+  detailed:
+    description:
+      return more in-depth settings for default subdictionary of
+      firewall_config. Matches the structure of the custom
+      subdictionary if this is active.
+    type: bool
+    required: false
+    default: false
+author: Brennan Paciorek (@BrennanPaciorek)
+"""
+
+RETURN = """
+changed:
+  description:
+    whether anything was changed, most likely false
+  type: bool
+  returned: always
+  sample: False
+firewall_config:
+  description:
+    firewall_config ansible fact,
+    system role sets this output to the
+    ansible fact named firewall_config
+  type: dict
+  returned: always
+  sample: {
+      "config": {
+          "default": {},
+          "custom": {},
+      },
+  }
+"""
+
+EXAMPLES = """
+# Run with no parameters to gather facts
+firewall_lib_facts:
+"""
+
+
+from ansible.module_utils.basic import AnsibleModule
+import os
+
+try:
+    import firewall.config
+
+    from firewall.client import FirewallClient
+
+    HAS_FIREWALLD = True
+except ImportError:
+    HAS_FIREWALLD = False
+
+try:
+    if HAS_FIREWALLD:
+        firewall.config.FIREWALLD_POLICIES
+
+    HAS_POLICIES = True
+except AttributeError:
+    HAS_POLICIES = False
+
+
+def config_to_dict(module):
+    detailed = module.params.get("detailed", False)
+    config = {}
+    defaults = {}
```

Comment:
Remove default sub-dictionary? I personally think we should keep it for better compatibility with use cases in the email. Aside from that, all we would need to do is decorate the README with warnings, both about cacheable (see tasks/main.yml review) and the size of `detailed` firewall_config ansible fact.

Reply:
I think we should keep it.